File: app/src/main/java/me/imzack/app/end/view/widget/EnhancedRecyclerView.java
package me.imzack.app.end.view.widget;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.util.AttributeSet;
import android.view.View;

public class EnhancedRecyclerView extends RecyclerView {

    private View mEmptyView;

    private AdapterDataObserver mAdapterDataObserver = new AdapterDataObserver() {
        @Override
        public void onChanged() {
            checkIfEmpty();
        }

        @Override
        public void onItemRangeInserted(int positionStart, int itemCount) {
            checkIfEmpty();
        }

        @Override
        public void onItemRangeRemoved(int positionStart, int itemCount) {
            checkIfEmpty();
        }
    };

    public EnhancedRecyclerView(Context context) {
        super(context);
    }

    public EnhancedRecyclerView(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public EnhancedRecyclerView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
    }

    private void checkIfEmpty() {
        if (mEmptyView == null || getAdapter() == null) return;
        boolean isAdapterEmpty = getAdapter().getItemCount() == 0;
        mEmptyView.setVisibility(isAdapterEmpty ? VISIBLE : GONE);
        setVisibility(isAdapterEmpty ? GONE : VISIBLE);
    }

    @Override
    public void setAdapter(Adapter adapter) {
        Adapter oldAdapter = getAdapter();
        if (oldAdapter != null) {
            oldAdapter.unregisterAdapterDataObserver(mAdapterDataObserver);
        }

        super.setAdapter(adapter);

        if (adapter != null) {
            adapter.registerAdapterDataObserver(mAdapterDataObserver);
        }

        checkIfEmpty();
    }

    public void setEmptyView(View emptyView) {
        mEmptyView = emptyView;
        checkIfEmpty();
    }
}

File: app/src/main/kotlin/me/imzack/app/end/view/widget/EnhancedRecyclerView.kt
package me.imzack.app.end.view.widget

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.util.AttributeSet
import android.view.View

class EnhancedRecyclerView : RecyclerView {

    private var mEmptyView: View? = null

    private val mAdapterDataObserver = object : RecyclerView.AdapterDataObserver() {
        override fun onChanged() {
            checkIfEmpty()
        }

        override fun onItemRangeInserted(positionStart: Int, itemCount: Int) {
            checkIfEmpty()
        }

        override fun onItemRangeRemoved(positionStart: Int, itemCount: Int) {
            checkIfEmpty()
        }
    }

    constructor(context: Context) : super(context)

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs)

    constructor(context: Context, attrs: AttributeSet, defStyle: Int) : super(context, attrs, defStyle)

    private fun checkIfEmpty() {
        if (mEmptyView == null || adapter == null) return
        val isAdapterEmpty = adapter.itemCount == 0
        mEmptyView!!.visibility = if (isAdapterEmpty) View.VISIBLE else View.GONE
        visibility = if (isAdapterEmpty) View.GONE else View.VISIBLE
    }

    override fun setAdapter(adapter: RecyclerView.Adapter<*>?) {
        val oldAdapter = getAdapter()
        oldAdapter?.unregisterAdapterDataObserver(mAdapterDataObserver)

        super.setAdapter(adapter)

        adapter?.registerAdapterDataObserver(mAdapterDataObserver)

        checkIfEmpty()
    }

    fun setEmptyView(emptyView: View) {
        mEmptyView = emptyView
        checkIfEmpty()
    }
}
